src/arch/riscv/pmp.cc

```diff
@@ -0,0 +1,223 @@
+    maxEntries(params.max_pmp),
+    numRules(0)
+{
+    for (int i=0; i < maxEntries; i++) {
+        PmpEntry entry;
+        pmpTable.emplace_back(entry);
+    }
+}
+
+Fault
+PMP::pmpCheck(const RequestPtr &req, BaseTLB::Mode mode,
+              RiscvISA::PrivilegeMode pmode, ThreadContext *tc)
+{
+    // First determine if pmp table should be consulted
+    if (!shouldCheckPMP(pmode, mode, tc))
+        return NoFault;
+
+    // An access should be successful if there are
+    // no rules defined yet or we are in M mode (based
+    // on specs v1.10)
+    if (numRules == 0 || (pmode == RiscvISA::PrivilegeMode::PRV_M))
+        return NoFault;
+
+    // all pmp entries need to be looked from the lowest to
+    // the highest number
+    int match_index = -1;
+    for (int i = 0; i < pmpTable.size(); i++) {
+        if (pmpTable[i].pmpAddr.contains(req->getPaddr())) {
+           //address matched
+            match_index = i;
+        }
+
+        if ((PMP_OFF != pmpGetAField(pmpTable[match_index].pmpCfg))
+                                            && (match_index > -1)) {
+            // check the RWX permissions from the pmp entry
+            uint8_t allowed_privs = PMP_READ | PMP_WRITE | PMP_EXEC;
+
+            // i is the index of pmp table which matched
+            allowed_privs &= pmpTable[match_index].pmpCfg;
+
+            // mode: R = 0, W = 1, X = 2
+            uint8_t mode_mask = 1 << mode;
+            if ((allowed_privs & mode_mask) > 0)
+                return NoFault;
+            else
+                return createAddrfault(req->getVaddr(), mode);
+        }
+    }
+    // if no entry matched and we are not in M mode return fault
+    return createAddrfault(req->getVaddr(), mode);
+}
+
+Fault
+PMP::createAddrfault(Addr vaddr, BaseTLB::Mode mode)
+{
+    RiscvISA::ExceptionCode code;
+    if (mode == BaseTLB::Read) {
+        code = RiscvISA::ExceptionCode::LOAD_ACCESS;
+    } else if (mode == BaseTLB::Write) {
+        code = RiscvISA::ExceptionCode::STORE_ACCESS;
+    } else {
+        code = RiscvISA::ExceptionCode::INST_ACCESS;
+    }
+    return std::make_shared<RiscvISA::AddressFault>(vaddr, code);
+}
+
+inline uint8_t
+PMP::pmpGetAField(uint8_t cfg)
+{
+    // to get a field from pmpcfg register
+    uint8_t a = cfg >> 3;
+    return a & 0x03;
+}
+
+
+void
+PMP::pmpUpdateCfg(uint32_t pmp_index, uint8_t this_cfg)
+{
+    pmpTable[pmp_index].pmpCfg = this_cfg;
+    pmpUpdateRule(pmp_index);
+
+}
+
+void
+PMP::pmpUpdateRule(uint32_t pmp_index)
+{
+    // In qemu, the rule is updated whenever
+    // pmpaddr/pmpcfg is written
+
+    numRules = 0;
+    Addr prevAddr = 0;
+
+    if (pmp_index >= 1) {
+        prevAddr = pmpTable[pmp_index - 1].rawAddr;
+    }
+
+    Addr this_addr = pmpTable[pmp_index].rawAddr;
+    uint8_t this_cfg = pmpTable[pmp_index].pmpCfg;
+    AddrRange this_range;
+
+    switch (pmpGetAField(this_cfg)) {
+      // checking the address matching mode of pmp entry
+      case PMP_OFF:
```

Comment:
I *think* these case statements should be at the same indentation as the switch. I could be wrong, though. I thought that the `public/private` keywords were the only thing we had at 2 spaces indented.

Reply:
I couldn't find any discussion on this in the style guide and looking into some other source files (e.g. isa.cc) it seems like there are 2 spaces indented cases that do not seem to have keywords in the case statement. Should I keep it as it is or just remove these spaces?